### ISSUE TITLE
Binary executables for Linux

### ISSUE THREAD

**Issue by The-Compiler** [The-Compiler/qutebrowser#1912]
Having Linux binaries would help us getting rid of older Qt version support (e.g. #1775) and bring QtWebEngine to people using Debian/Ubuntu.

There are various approaches:

- Simply telling people to install `PyQt5` from PyPI
- [PyInstaller](http://www.pyinstaller.org/) (like for OS X, and hopefully soon Windows)
- [AppImage](http://appimage.org/) (needed `libselinux` from the AUR on Arch... universal? Bleh.
- [Flatpak](http://flatpak.org/)
- [Snap](http://snapcraft.io/)
- [Docker](https://www.docker.com/)

**Comment by probonopd** [comment]
> AppImage (needed libselinux from the AUR on Arch... universal? Bleh.)

Please explain. AppImage per se does not require libselinux and particularly not from the AUR on Arch. (Some particular app that comes inside an AppImage might; then it could bundle it inside the AppImage if so desired.)

**Comment by The-Compiler** [comment]
I tried the Subsurface demo AppImage linked on http://appimage.org/ ([this file](https://bintray.com/probono/AppImages/download_file?file_path=Subsurface-4.5.1.1445-x86_64.AppImage)) and without `libselinux` installed, I get:

```
./Subsurface-4.5.1.1445-x86_64.AppImage
/tmp/.mount_ecIaHT
/tmp/.mount_ecIaHT/usr/bin/subsurface.wrapper
/tmp/.mount_ecIaHT/subsurface.desktop
/tmp/.mount_ecIaHT/usr/bin/subsurface: error while loading shared libraries: libselinux.so.1: cannot open shared object file: No such file or directory
```

Sorry - my bad for assuming it was AppImage in general requiring that. However I can't see what subsurface would use selinux for. I tried the Krita AppImage instead, and that seems to run fine. Still, not the best first impression for the official demo AppImage :wink:

**Comment by probonopd** [comment]
Thanks for pointing this out, you are right, this is not how it should be. Will follow up with the Subsurface team.

**Comment by The-Compiler** [comment]
So I just played around with PyInstaller a bit since we're using that on OS X and soon Windows too.

I could get an image built on Archlinux run on Arch with a few missing files copied:

```
cp ~/proj/qutebrowser/git/.tox/pyinstaller/lib/python3.6/site-packages/PyQt5/Qt/libexec/QtWebEngineProcess . 
cp -R ~/proj/qutebrowser/git/.tox/pyinstaller/lib/python3.6/site-packages/PyQt5/Qt/resources/* .
cp -R ~/proj/qutebrowser/git/.tox/pyinstaller/lib/python3.6/site-packages/PyQt5/Qt/translations/qtwebengine_locales .
cp -R ~/proj/qutebrowser/git/.tox/pyinstaller/lib/python3.6/site-packages/PyQt5/Qt/plugins/* qt5_plugins/
```

Then I tried building an image with PyQt5 from PyPI on Ubuntu Trusty, but first needed Python 3.6.

I installed pyenv and installed it using `env PYTHON_CONFIGURE_OPTS="--enable-shared` pyenv install 3.6.0`

However, I never got that image to run anywhere.

On Trusty:

```
ImportError: libsmime3.so: cannot open shared object file: No such file or directory
```

On Arch:

```
ImportError: /usr/lib/libGL.so.1: undefined symbol: drmFreeDevice
```

After deleting `libdrm.so.2` as per [this SO post](http://stackoverflow.com/questions/40998027/genymotion-genymotion-symbol-lookup-error-usr-lib64-libgl-so-1-undefined) I just get a segfault here:

```
  File "qutebrowser/misc/earlyinit.py", line 292 in check_ssl_support
```

On Debian testing there are also various gl and SSL errors (though it seems those happen with PyQt5 from PyPI in general)


I'll probably try AppImage next, but seeing that Debian Stretch has QtWebEngine packaged and PyQt5 from PyPI can be used relatively easily on other distributions, this is getting less important.

**Comment by sanjayankur31** [comment]
I've set up a COPR repository here for Fedora users - but this is tracking the Git HEAD and not stable releases for the moment. I'll look into Flatpak when I have some time - would be really helpful!

https://copr.fedorainfracloud.org/coprs/ankursinha/qutebrowser/

Thanks for all the awesome work! Looking forward to everything that's planned!

**Comment by tidux** [comment]
Adding a +1 for Flatpak over appimage for the fact that you can run standard Flatpaks on non-glibc distributions like Alpine.  This gives the best universality for a single build.

**Comment by probonopd** [comment]
@tidux but it comes at the expense that the Flatpak package itself and a runtime must be installed on the system before you can run the application.

**Comment by tidux** [comment]
Flatpak local installs can be done per user with no root rights required. 

On July 6, 2017 4:05:50 PM PDT, probonopd <notifications@github.com> wrote:
>@tidux but it comes at the expense that the Flatpak package itself and
>a runtime must be installed on the system before you can run the
>application.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/qutebrowser/qutebrowser/issues/1912#issuecomment-313542956

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by tidux** [comment]
Further update for Flatpak: https://github.com/flathub/flathub is providing a place for people to upload flatpaks without needing to host their own server infrastructure.  @probonopd every modern distro has Flatpak.  Even SLACKWARE has Flatpak as a Slackbuild, and they were shipping LILO into the 2010s.  The main desktoppy distros install it by default these days, and Nautilus has filetype integration.

**Comment by The-Compiler** [comment]
FWIW hosting isn't an issue for qutebrowser, I do already have my own server infrastructure (which e.g. also hosts qutebrowser.org).

**Comment by probonopd** [comment]
> @probonopd every modern distro has Flatpak.

In your dreams. The day Debian and Ubuntu have Flatpak working out of the box in the default installation I will buy you a beer. Seriously.

**Comment by tidux** [comment]
I take payment in [P-51 Coconut Porter.](https://www.beeradvocate.com/beer/profile/24029/102161/)

```
$ cat /etc/debian_version 
9.1
$ flatpak --version
Flatpak 0.8.5
$ dpkg --get-selections | grep flatpak
flatpak                                         install
flatpak-builder                                 install

```

No configuration is needed and the packages are in debian-main, including the GNOME Software integration plugin (gnome-software-plugin-flatpak) as a suggested dependency of gnome-software.  I haven't used the Tasksel auto-built desktop in years, so I can't say as to what's in the "default install" of Debian.  Ubuntu 17.04 will have Flatpak integration installed by default.  http://www.ubuntubuzz.com/2017/03/a-short-review-of-ubuntu-gnome-1704-beta-1.html

**Comment by haasn** [comment]
> Adding a +1 for Flatpak over appimage for the fact that you can run standard Flatpaks on non-glibc distributions like Alpine. This gives the best universality for a single build.

Yeah but you can run an appimage on any distro without having to install some horrible runtime first. Just download and executable, which is exactly what a “portable executable” format should be.

**Comment by The-Compiler** [comment]
I really don't care much about the whole "AppImage vs. Flatpak" thing so far FWIW - I'll take whatever is easier to actually get to work, and works on things like Ubuntu 14.04 or Debian Jessie.

**Comment by KurtPfeifle** [comment]
Any chance to get a *`qutebrowser.AppImage`* any time soon?

Would you accept a pull request if it helps you to automatically build it with Travis CI for each release (or even for each commit)?

----

<sup>(I mean for a developer who already builds .dmg for his Mac users, going for .AppImage packages to distribute to his Linux users shouldn't be too far a mental stretch, no?)</sup>

**Comment by The-Compiler** [comment]
Building it automatically isn't really the goal for now - the first step would be getting a manual build to run properly. So far this hasn't happened, mainly because the situation with library versions and graphic drivers is quite a lot simpler on Windows/macOS than it is on Linux. Help with that would certainly be welcome, though note that you can use a PyQt install from PyPI (e.g. via tox, see the install instructions) if you have a Python 3.5+ on x86_64.

**Comment by KurtPfeifle** [comment]
> *"situation with library versions and graphic drivers is quite a lot simpler on Windows/macOS than it is on Linux. Help with that would certainly be welcome"*

No promise, but I'll see what I can do myself (or what other help I can organize).

**Comment by KurtPfeifle** [comment]
Oh, and meanwhile it would be *nice if you could add a short list of instructions about how to install the `.deb` package*. Here:

* https://github.com/qutebrowser/qutebrowser/releases

I had fetched the .deb, but it would not properly install on Debian Stretch:

<pre>
Selecting previously unselected package qutebrowser.
(Reading database ... 476521 files and directories currently installed.)
Preparing to unpack qutebrowser_1.0.4-1_all.deb ...
Unpacking qutebrowser (1.0.4-1) ...
dpkg: dependency problems prevent configuration of qutebrowser:
 qutebrowser depends on python3-attr; however:
  Package python3-attr is not installed.
 qutebrowser depends on python3-pyqt5.qtsql; however:
  Package python3-pyqt5.qtsql is not installed.
 qutebrowser depends on python3-pyqt5.qtopengl; however:
  Package python3-pyqt5.qtopengl is not installed.
 qutebrowser depends on python3-pyqt5.qtquick; however:
  Package python3-pyqt5.qtquick is not installed.
 qutebrowser depends on python3-pyqt5.qtwebengine (>= 5.7); however:
  Package python3-pyqt5.qtwebengine is not installed.
 qutebrowser depends on libqt5webenginecore5 (>= 5.7.1); however:
  Package libqt5webenginecore5 is not installed.
 qutebrowser depends on python3-pypeg2; however:
  Package python3-pypeg2 is not installed.
 qutebrowser depends on python3-ruamel.yaml; however:
  Package python3-ruamel.yaml is not installed.
 qutebrowser depends on python3-jinja2; however:
  Package python3-jinja2 is not installed.

dpkg: error processing package qutebrowser (--install):
 dependency problems - leaving unconfigured
Processing triggers for desktop-file-utils (0.23-1) ...
Processing triggers for mime-support (3.60) ...
Processing triggers for hicolor-icon-theme (0.15-1) ...
Processing triggers for man-db (2.7.6.1-2) ...
Processing triggers for menu (2.1.47+b1) ...
Errors were encountered while processing:
 qutebrowser
</pre>

And then:

<pre>
kp@~>  sudo apt-get install \
                    python3-attr python3-pyqt5.qtsql \
                    python3-pyqt5.qtopengl python3-pyqt5.qtquick \
                    python3-pyqt5.qtwebengine libqt5webenginecore5 \
                    python3-pypeg2 python3-ruamel.yaml python3-jinja2
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Package python3-pypeg2 is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

E: Package 'python3-pypeg2' has no installation candidate
</pre>

Turns out, even a *`.deb`* in the end is of not-so-great benefit for users if it (or its dependencies) do not smoothly install on Debian Stretch.

*That* problem is what brought me originally here, to search for "AppImage"   :-)

**Comment by The-Compiler** [comment]
Please check the [install docs](https://github.com/qutebrowser/qutebrowser/blob/master/doc/install.asciidoc).

**Comment by KurtPfeifle** [comment]
> *"Please check the install docs."*

Thanks. Will do.

However, to at least **EASE** the situation for users who are willing to jump through the loops of installing a *`.deb`* which does not work out of the box, could you please link to the install docs PROMINENTLY from that download page?

Debian users are a bit spoiled (at least I am): when I see a *`.deb`* with no extra explanation, I assume it will install without extra gymnastics...

**Comment by The-Compiler** [comment]
No, because I'd have to link to the install docs in every release changelog, and I'd rather not do that (because it has nothing to do with the changelog).

Especially if installation doesn't work, I expect people to find an install documentation linked prominently in the readme and in the header menu on qutebrowser.org.

**Comment by KurtPfeifle** [comment]
You are quite stubborn, aren't you? :-)
<sup>(No offense intended -- stubbornness on many occasions is a quite good thing.)</sup>

Ok, let me try one more time:

* Your install hints are quite difficult to find. Can you please put a little, little bit of effort in, so they are better visible? Suggestions:
    1. Make the link in the `README.asciidoc` more PROMINENT. Give it a headline of its own?
    1. Elevate the location of the `INSTALL.asciidoc` from the *docs/* subdir one level higher?

* Still -- you *could* link the install doc manually from the release changelog (yes, every time). Or make it automatically appear on that GitHub download page (I would be surprised if there weren't a way to do it). That you don't *want* it, is a different thing...

And it may not have anything to do with the CHANGELOG, yes. But it has *everything* to do with the installation of the linked downloads, especially as it does not work as smoothly as one would expect when seeing a *`.deb`* in the shop window!  
<sub>(With your argument, you could also say you should not even have the changelog located there, because it has "nothing to do with the downloads"...)</sub>

Maybe you don't **want** to make it as easy as possible for your users to use the software you should be proud of having created? Maybe this project is mainly for this own little private fun you have and enjoy when solving technical problems and finding solutions? Well, that would be fair enough, even if disappointing for users...

I still want to thank you for qutebrowser. I have it finally running. Though on Mac now. It is really cute! Thanks again.
👍 <3

**Comment by The-Compiler** [comment]
> Your install hints are quite difficult to find.

I still don't follow. It's literally in the first section in the readme after the screenshot ("Downloads"), and in the list of other documentation again.

I'll add a link near the "Releases" at the very top of the readme though.

> Elevate the location of the INSTALL.asciidoc from the docs/ subdir one level higher?

Which probably makes it harder to find (because a `docs/` subfolder is exactly where I'd expect to find... well.. docs :wink:) and clutters the root directory more.

> Or make it automatically appear on that GitHub download page (I would be surprised if there weren't a way to do it).

Well, there isn't.

> With your argument, you could also say you should not even have the changelog located there, because it has "nothing to do with the downloads"... 

Except that it's a *description of that release*, which is exactly what a changelog is.

> Maybe you don't want to make it as easy as possible for your users to use the software you should be proud of having created? Maybe this project is mainly for this own little private fun you have and enjoy when solving technical problems and finding solutions? Well, that would be fair enough, even if disappointing for users...

Please stop jumping to (entirely wrong) conclusions. We're talking about some 700 downloads of the .deb for each release, and you're the first one to complain about not finding the installation docs.

That being said, qutebrowser is soon going to land in the Debian testing repos anyways (v1.0.4 landed in unstable today), which makes things more straightforward.

Anyways - if you insist on continuing this discussion, please take it to [email](mailto:mail@qutebrowser.org) or [IRC](https://github.com/qutebrowser/qutebrowser#getting-help). This has literally nothing to do with AppImages.

**Comment by The-Compiler** [comment]
FYI: After @KurtPfeifle suggesting it over email, I deleted the offtopic comments here.

**Comment by The-Compiler** [comment]
I got a mail from @torsava, citing it here with his permission:

> I was trying out flatpak, and took a swing at packaging qutebrowser. I 
> managed to get it working and solved some issues (udev, graphics). It's 
> not ideal (more in README) and I'm not going to pursue it further, 
> nevertheless, if anyone would like to give it a try, here's what I got:
> 
> https://github.com/torsava/flatpak-qutebrowser

**Comment by glatzor** [comment]
I created a new flatpak repository:

https://github.com/glatzor/flathub/tree/org.qutebrowser.qutebrowser

@The-Compiler  If you are ok with the quality I would like to submit it to flathub.

**Comment by The-Compiler** [comment]
I've just tried using PyInstaller on Linux again, and with a [quick fix](https://github.com/pyinstaller/pyinstaller/pull/3758), at least a binary built on Archlinux works there. If we can build with that on some old Ubuntu, that might be a nice solution.

If that works, that'd be my preferred solution, because we already use PyInstaller for Windows/macOS.

**Comment by curioussavage** [comment]
I just installed the flatpak here https://github.com/flathub/flathub/pull/814 It works great!

**Comment by Aledosim** [comment]
Any news on this? I'm a Debian based distro user, I'd love to have some up to date AppImage and watch videos with proprietary codecs

**Comment by The-Compiler** [comment]
Not that I'm aware of.

**Comment by KIC** [comment]
I was wondering if we could use pyinstaller with this project and found its partially used already and also found this issue. So what is the status here, since this issue is still "open". I think it would be a big benefit just dropping a binary to the note KDE linux folks instead of building from source. Is there a spec file already which works with qt6?

btw. the flatpack version is 2.5.4
```
flatpak search org.qutebrowser.qutebrowser
Name                 Description                                                                                   Application ID                                   Version          Branch          Remotes
Userscripts          qutebrowser userscripts supporting Python modules, shared libraries, and executables          org.qutebrowser.qutebrowser.Userscripts                           stable          flathub
qutebrowser          A keyboard-driven web browser                                                                 org.qutebrowser.qutebrowser                      2.5.4            stable          flathub
```
Is flakpack still a thing? to me it looks like dead.

**Comment by The-Compiler** [comment]
[There is a .spec](https://github.com/qutebrowser/qutebrowser/blob/main/misc/qutebrowser.spec), it's what it's used to build the Windows/macOS releases. You can probably use it to build a Linux binary, but honestly I don't really see the point over just [using a virtualenv install](https://github.com/qutebrowser/qutebrowser/blob/main/doc/install.asciidoc#installing-qutebrowser-with-virtualenv) which gives you the same thing but with more flexibility and less hacks.

I don't know what you mean by "building from source", there isn't really much in qutebrowser to build.

The qutebrowser Flatpak is indeed unmaintained (it was never maintained by any qutebrowser maintainer). Flatpak in general still seems very much a thing.
